### PR TITLE
Integrate modules for sampling

### PR DESCRIPTION
**Pull request by CoreLeader**
# Issues to be solve

这个分支目前来说并不是一个可以直接合并到dev分支的分支，主要是有一些问题需要和大家商量着解决。主要原因是在model里面和LP，在编程上有一些出入。

每一个model里面实际有两层，外层是有Auto前缀的，用于根据HP实例化内层的继承torch.nn.Module的具体实现（e.g. GCN, GAT, GraphSAGE）。之前讨论过，对于实际实现model的内层类需要拆分成encode和decode两个阶段，LP也需要拆分成这两个阶段，sampling也是需要这两个阶段，甚至需要对外暴露model的每一层。

**现在主要问题就是LP的decode和sampling的decode是不同的**，但是昨天merge的LP的decode功能直接命名为了`decode(self, ...)`，这样可能是不是不太好，因为一个model不仅仅用于LP，也用于已经实现的classification。我建议是不是可以改成`link_prediction_decode(self, ...)`或者`lp_decode(self, ...)`一类的函数名称，带上task的前缀或后缀，我在sampling的开发分支上也可以相应修改为`cls_decode(self, x)`一类的函数名称，具体就看我们怎么样约定decode功能函数的命名规范。

由于昨天lp分支已经merge到了dev分支，我建议负责LP的昊阳师兄和其他的师兄先改一改`gcn.py`和`graphsage.py`中decode方法的名称merge到dev分支，之后我再Rebase，再修改sampling功能的PR。

对于外层的有Auto前缀的注册入MODEL_DICT的类继承BaseModel和ClassificationModel我没有意见，问题不大。

# Sampling Performance Comparison

## Node-wise Sampling (GraphSAGE)

1. Reddit
* AutoGL

Val: 0.9505

Test: 0.9475

* Baseline (PyG example)

Val: 0.9509

Test: 0.9501

## Layer-wise Sampling (LADIES)

Sampling size: 512

1. Cora
* AutoGL

Val: 0.788

Test: 0.7950

* Baseline (original literature)

Test: 78.3% (+-1.6%)

2. Citeseer
* AutoGL

Val: 0.694

Test: 0.6680

* Baseline (original literature)

Test: 64.3% (+-2.4%)

3. PubMed
* AutoGL

Val: 0.802

Test: 0.7590

* Baseline (original literature)

Test: 75.9% (+-1.1%)

4. Reddit
* AutoGL

Test: 91.75%

* Baseline (original literature)

Test: 92.8% (+-1.6%)

## Subgraph-wise Sampling （GraphSAINT）

1. Flickr
* AutoGL

Val: 0.510129

Test: 0.5071

* Original Literature

Test: 0.511

* AutoGL的GraphSAINT模型参考了论文作者所提供的源代码（[https://github.com/GraphSAINT/GraphSAINT](https://github.com/GraphSAINT/GraphSAINT)），尽可能还原了作者的模型，故将原论文作为Baseline进行对比。

### CODE REVIEW

**Comment by Frozenmad** [comment]
@lihy96 Can you revise the related `decode()` method to `lp_decode()` and pr again?

**Comment by Frozenmad** [comment]
@CoreLeader The method names are now changed to `lp_decode` from PR #32. You can go ahead.

File: autogl/solver/utils.py [hunk 1]
```diff
@@ -1,23 +1,23 @@
 """
```
Comment (Frozenmad) [inline autogl/solver/utils.py:1]: First, thanks for fixing typos and bringing new output format. But remember, not everyone is happy when his/her designs or codes are modified without discussion or at least notification. Next time when you want to bring new features or change logic flow, you **should** first discuss with the one associated with the codes/classes.

There are also some conflicts with my design earlier, please find my comments below.

File: autogl/solver/utils.py [hunk 2]
```diff
@@ -49,64 +71,74 @@ def set_major_field(self, field) -> None:
...
+        if field in self.__fields:
+            self.__major_field = field
         else:
             LOGGER.warning(
-                "do not find major field %s in current leaderboard, will ignore.", field
+                "do not find major field %s in the current LeaderBoard, will ignore.", field
             )
 
-    def insert_model_performance(self, name, performance) -> None:
+    def add_performance(
```
Comment (Frozenmad) [inline autogl/solver/utils.py:81]: The LeaderBoard is designed in dictionary format, which means: the models and rows of dataframe are 1-to-1 pairs. The names of models are **not allowed** to be the same. In your new design, at least names + representations are **not allowed** to be the same.
Please change the logic back.

File: autogl/solver/utils.py [hunk 3]
```diff
@@ -49,64 +71,74 @@ def set_major_field(self, field) -> None:
...
+            __representation: str = representation
         else:
-            LOGGER.warning(
-                "model already in the leaderboard, will override current result."
-            )
-            self.remove_model_performance(name)
-            self.insert_model_performance(name, performance)
+            raise TypeError
 
-    def remove_model_performance(self, name) -> None:
```
Comment (Frozenmad) [inline autogl/solver/utils.py:89]: For completeness of this object class. Please add the logic back.

File: autogl/solver/utils.py [hunk 4]
```diff
@@ -122,34 +154,67 @@ def get_best_model(self, index=0) -> str:
         name: `str`
             The name/identifier of the required model.
         """
-        sorted_df = self.perform_dict.sort_values(
-            by=self.major_field, ascending=not self.is_higher_better[self.major_field]
+        sorted_performance_df = self.__performance_data_frame.sort_values(
+            self.__major_field,
+            ascending=not (
+                dict(zip(self.__fields, self.__is_higher_better))[self.__major_field]
```
Comment (Frozenmad) [inline autogl/solver/utils.py:160]: Will there be cases when `self.__major_filed` not in `self.__fields`?

File: autogl/solver/utils.py [hunk 5]
```diff
@@ -122,34 +154,67 @@ def get_best_model(self, index=0) -> str:
...
+            _columns.remove("name")
+            _columns.insert(0, "name")
+            maxcolwidths.append(40)
+        self.__performance_data_frame = self.__performance_data_frame[_columns]
+
+        sorted_performance_df: pd.DataFrame = self.__performance_data_frame.sort_values(
+            self.__major_field,
+            ascending=not (
+                dict(zip(self.__fields, self.__is_higher_better))[self.__major_field]
+                if self.__major_field in self.__fields else True
```
Comment (Frozenmad) [inline autogl/solver/utils.py:205]: Will there be cases when `self.__major_filed` not in `self.__fields`?

File: autogl/solver/utils.py [hunk 6]
```diff
@@ -122,34 +154,67 @@ def get_best_model(self, index=0) -> str:
...
+        '''
+        reindex self.__performance_data_frame
+        to ensure the columns of name and representation are in left-side of the data frame
+        '''
+        _columns = self.__performance_data_frame.columns.tolist()
+        maxcolwidths: _typing.List[_typing.Optional[int]] = []
+        if "representation" in _columns:
+            _columns.remove("representation")
+            _columns.insert(0, "representation")
+            maxcolwidths.append(40)
```
Comment (Frozenmad) [inline autogl/solver/utils.py:194]: I _recommend_ the `maxcolwidths` be set as a controllable attribute for this object class, or an argument for this method.

File: autogl/solver/utils.py [hunk 7]
```diff
@@ -122,34 +154,67 @@ def get_best_model(self, index=0) -> str:
...
+        _columns = sorted_performance_df.columns.tolist()
+        maxcolwidths.extend([None for _ in range(len(_columns) - len(maxcolwidths))])
         print(
-            self.perform_dict.sort_values(
-                by=self.major_field,
-                ascending=not self.is_higher_better[self.major_field],
-            ).head(top_k)
+            tabulate(
+                list(zip(*[sorted_performance_df[column] for column in _columns])),
+                headers=_columns, tablefmt="grid"
```
Comment (Frozenmad) [inline autogl/solver/utils.py:216]: wish to see the new output :)

**Review by Frozenmad — COMMENTED** [review]
Thanks for your constructive contributions. The sampling/large-scale attribution will become an essential and important part of our repository! However, I notice that there are many changes to original source codes. We need to double-check the performances (not just for sampling, but for tasks already supported) before merging this branch.

Also, please be aware:
- Many times one functionality can be implemented through different ways. But if there already exists one implementation, we should **avoid re-implementing/re-factoring unless necessary**.
- If you need new features or want to update some features that already exist, please **first** contact with the **developers who are responsible for corresponding features** and **let them do the work**. First, you don't have to write codes that are supposed to be other's responsibility. Second, not everyone is happy when you change the his/her codes without any discussion or even notification, since it may imply some kinds of disrespect.


There are also some detailed comments. Please find them below.

File: autogl/module/hpo/base.py [hunk 8]
```diff
@@ -30,7 +30,7 @@ def get_depended_para(name):
             raise WrongDependedParameterError("The depended parameter does not exist.")
 
         for para in config:
-            if para["type"] == "NUMERICAL_LIST" and para.get("cutPara", None):
+            if para["type"] in ("NUMERICAL_LIST", "CATEGORICAL_LIST") and para.get("cutPara", None):
```
[the remark this answers is not in the document]
Comment (general502570) [reply autogl/module/hpo/base.py:33]: It's OK

**Comment by CoreLeader** [comment]
I am really appreciated for the previous reviews and comments proposed by other people, not only for this pull request but also for the way of collaborating with other developers. Now the involved modules have been revised and tested, somes details and notices are as follows.
### Potential influences for other modules
1. Now the encoding method for link prediction is named as `lp_encode(self, data)`.
The reason is that the use approaches of models are diverse for classification and link prediction, and thus now there are four optional methods in model, e.g. `lp_encode`, `lp_decode`, `cls_encode`, `cls_decode`.
For a specific model which inherits `ClassificationSupportedSequentialModel`, classification trainer would call `cls_encode` and `cls_decode` methods, otherwise the `forward` (i.e. `__call__`) method would be called according to the convention of PyTorch.
2. **Please do not set *dropout* hp in YAML configuration file if dropout mechanism shall not adopted**
For current head commit of `dev` branch, I notice that dropout mechnism is disabled by commenting out.
https://github.com/THUMNLab/AutoGL/blob/fbd420e8e51098dfdcd179e75bd8dde9b19f17bc/autogl/module/model/gcn.py#L93
https://github.com/THUMNLab/AutoGL/blob/fbd420e8e51098dfdcd179e75bd8dde9b19f17bc/autogl/module/model/graphsage.py#L180
https://github.com/THUMNLab/AutoGL/blob/fbd420e8e51098dfdcd179e75bd8dde9b19f17bc/autogl/module/model/gat.py#L104
These may be shortcuts for implementing Link Prediction, however, dropout mechanism is essential for classification to improve generalization capability. Therefore, if dropout mechanism shall not adopted for Link Prediction, please do not include *dropout* hyper-parameter in HP space, the corresponding config files `configs/lp_*.yml` may need to be revised. (Don't worry, the wrapper classes `AutoGCN` and `AutoSAGE` will set dropout as `None` for corresponding inner model when the dropout HP is not Provided. I have not verify whether other models including GAT have simillar guarantee yet, if needed, I am willing to revise other models)
> Furthermore, for GCN and GraphSAGE, the components for one layer, i.e. Convolution, Activation, and optional dropout, have been wrapped into a wrapper class, named as `_XXXLayer`, because that different layers have different dropout probabilities, is significant for large-scale classification. Besides, a layer-wise evaluation technique is a practical evaluation approach for lar-scale rerpresentation encoding in evaluation time. See the source codes in `graphsage.py` and `gcn.py` for more details. In addition, the `graph_sage.py` has been removed.

If someone have some different ideas for the models, please be aware of the significance and process  for layer-wise evaluation technique in advance, and modify the models accordingly.

3. `leaderboard` -> `LeaderBoard`
The inappropriate modifications for leader board have been reverted, now the functionalities of leader board are consistent with original design, except the `show()` method.

**Comment by CoreLeader** [comment]
Even thogh their are still some conflicts for VCS, please focus on the modifications. Once the majority of modifications is approved, I will merge from current `dev` branch and make it ready for merge into `dev`.

**Comment by Frozenmad** [comment]
Thanks for your contribution! I've gone through all the changes and have no further comments. 
Before merging this branch, we are planning to run some key results on the final merged branch to see whether the performance of nodeclf, graphclf and lp is kept. Can you resolve the conflicts (by merge dev into your branch, or creating a new branch) and make it ready for the test?
